File: backend/libexecution/libdb.ml
open Core_kernel
open Lib
open Runtime
open Types.RuntimeT

let fns : Lib.shortfn list =
  [ { pns = ["DB::insert"]
    ; ins = []
    ; p = [par "val" TObj; par "table" TDB]
    ; r = TObj
    ; d = "DEPRECATED: Insert `val` into `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::delete"]
    ; ins = []
    ; p = [par "value" TObj; par "table" TDB]
    ; r = TNull
    ; d = "DEPRECATED: Delete `value` from `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::deleteAll"]
    ; ins = []
    ; p = [par "table" TDB]
    ; r = TNull
    ; d = "DEPRECATED: Delete everything from `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::update"]
    ; ins = []
    ; p = [par "value" TObj; par "table" TDB]
    ; r = TNull
    ; d = "DEPRECATED: Update `table` value which has the same ID as `value`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::fetchBy"]
    ; ins = []
    ; p = [par "value" TAny; par "field" TStr; par "table" TDB]
    ; r = TList
    ; d =
        "DEPRECATED: Fetch all the values in `table` whose `field` is `value`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::fetchOneBy"]
    ; ins = []
    ; p = [par "value" TAny; par "field" TStr; par "table" TDB]
    ; r = TAny
    ; d =
        "DEPRECATED: Fetch exactly one value in `table` whose `field` is `value`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::fetchByMany"]
    ; ins = []
    ; p = [par "spec" TObj; par "table" TDB]
    ; r = TList
    ; d =
        "DEPRECATED: Fetch all the values from `table` which have the same fields and values that `spec` has"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::fetchOneByMany"]
    ; ins = []
    ; p = [par "spec" TObj; par "table" TDB]
    ; r = TAny
    ; d =
        "DEPRECATED: Fetch exactly one value from `table`, which have the samea fields and values that `spec` has"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::fetchAll"]
    ; ins = []
    ; p = [par "table" TDB]
    ; r = TList
    ; d = "DEPRECATED: Fetch all the values in `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::keys"]
    ; ins = []
    ; p = [par "table" TDB]
    ; r = TList
    ; d = "DEPRECATED: Fetch all the keys in `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true }
  ; { pns = ["DB::schema"]
    ; ins = []
    ; p = [par "table" TDB]
    ; r = TObj
    ; d = "DEPRECATED: Fetch all the values in `table`"
    ; f = NotClientAvailable
    ; ps = false
    ; dep = true } ]

File: fsharp-backend/src/LibBackend/StdLib/LibDb.fs
module LibBackend.StdLib.LibDB

open Prelude
open LibExecution.RuntimeTypes

let fn = FQFnName.stdlibName

let err (str : string) = Value(Dval.errStr str)

let removedFunction = LibExecution.Errors.removedFunction

let varA = TVariable "a"
let dbType = TDB varA

let fns : List<BuiltInFn> =
  [ { name = fn "DB" "insert" 0
      parameters = [ Param.make "val" varA ""; Param.make "table" dbType "" ]
      returnType = varA
      description = "Insert `val` into `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "delete" 0
      parameters = [ Param.make "value" varA ""; Param.make "table" dbType "" ]
      returnType = TNull
      description = "Delete `value` from `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "deleteAll" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TNull
      description = "Delete everything from `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "update" 0
      parameters = [ Param.make "value" varA ""; Param.make "table" dbType "" ]
      returnType = TNull
      description = "Update `table` value which has the same ID as `value`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "fetchBy" 0
      parameters =
        [ Param.make "value" TAny ""
          Param.make "field" TStr ""
          Param.make "table" dbType "" ]
      returnType = TList TAny
      description = "Fetch all the values in `table` whose `field` is `value`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "fetchOneBy" 0
      parameters =
        [ Param.make "value" TAny ""
          Param.make "field" TStr ""
          Param.make "table" dbType "" ]
      returnType = TAny
      description = "Fetch exactly one value in `table` whose `field` is `value`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "fetchByMany" 0
      parameters = [ Param.make "spec" varA ""; Param.make "table" dbType "" ]
      returnType = TList TAny
      description =
        "Fetch all the values from `table` which have the same fields and values that `spec` has"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "fetchOneByMany" 0
      parameters = [ Param.make "spec" varA ""; Param.make "table" dbType "" ]
      returnType = TAny
      description =
        "Fetch exactly one value from `table`, which have the same fields and values that `spec` has"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "fetchAll" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TList TAny
      description = "Fetch all the values in `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "keys" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = TList TAny
      description = "Fetch all the keys in `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) }
    { name = fn "DB" "schema" 0
      parameters = [ Param.make "table" dbType "" ]
      returnType = varA
      description = "Fetch all the values in `table`"
      fn = InProcess removedFunction
      sqlSpec = NotQueryable
      previewable = Impure
      deprecated = ReplacedBy(fn "" "" 0) } ]
